docs/coding-guidelines/typescript.md

```diff
@@ -0,0 +1,300 @@
+TypeScript Coding Guidelines
```

Comment:
This doc is absolutely fantastic, @dmsnell – it's everything I suggested in the vaguest way possible as being really helpful. The `Avoid / Prefer` model is super helpful. <3